openedx_blockstore_relay/transfer_data.py

```diff
@@ -0,0 +1,257 @@
+
+    if block_type == 'vertical':
+        transfer_obj = TransferUnit(block_key, bundle_uuid, asset_root_url)
+    elif block_type in ['sequential', 'chapter', 'course']:
+        # TODO: handle full descendent tree to support these block types.
+        raise NotImplementedError("Block type not supported: %s" % block_type)
+    else:
+        transfer_obj = TransferBlock(block_key, bundle_uuid, asset_root_url)
+
+    transfer_obj.transfer_block_to_bundle()
+
+
+class TransferBlock(object):
+    """
+    Transfers a single XBlock and its related files from Open edX to Blockstore.
+
+    Uploads the sharable interface of the Bundle in a manifest bundle.json file.
+    """
+
+    BUNDLE_VERSION = 1
+
+    def __init__(self, block_key, bundle_uuid, asset_root_url=None):
+        """
+        Store the block_key (UsageKey) and bundle_uuid (UUID).
+        """
+        super(TransferBlock, self).__init__()
+        self.block_key = block_key
+        self.bundle_uuid = bundle_uuid
+        self.bundle_files_url = urljoin(settings.BLOCKSTORE_API_URL,
+                                        '/'.join(['bundles', str(self.bundle_uuid), 'files', '']))
+        self.manifest = {
+            'meta': {'version': self.BUNDLE_VERSION},
+            'type': self.olx_type(self.block_key.block_type),
+            'assets': [],
+            'components': [],
+            'dependencies': [],
+        }
+        self.assets = []
+        self.asset_root_url = asset_root_url or settings.LMS_ROOT_URL
+
+    def transfer_block_to_bundle(self):
+        """
+        Upload block-related files to the Bundle.
+        """
+        LOG.debug('Transfer Unit %s to Bundle <%s>', self.block_key, self.bundle_uuid)
+        self.upload_block()
+        self.upload_assets()
+        self.upload_manifest()
+
+    def upload_block(self):
+        """
+        Upload a block and its related components to the bundle.
+        """
+        LOG.debug('Transfer %s to Bundle <%s>', self.block_key, self.bundle_uuid)
+        filesystem = MemoryFS()
+        block = self.get_block()
+        olx_node = block.definition_to_xml(filesystem)
+        self.upload_components(filesystem)
+        self.upload_olx(olx_node, block, store_component=True)
+
+    def get_block(self):
+        """
+        Retrieve the block from the modulestore.
+
+        Raises modulestore.exceptions.ItemNotFoundError if the block is not found.
+        """
+        from xmodule.modulestore.django import modulestore  # pylint: disable=import-error
+        store = modulestore()
+        return store.get_item(self.block_key)
+
+    @staticmethod
+    def olx_type(block_type):
+        """
+        Return the content_type to set for the given block_key.
+        """
+        return 'olx/{}'.format(block_type)
+
+    def upload_components(self, filesystem):
+        """
+        Upload any files found in the filesystem.
+        """
+        for item in filesystem.walk():
```

Comment:
Can you explain what is happening here? Also what is being uploaded?

Reply:
I've renamed this method to make it clearer, and expanded on the comments.  Does that help?

cf https://github.com/open-craft/openedx-blockstore-relay/pull/1/commits/9e4e18121e377c293be33576637fe5f1f07fdf53#diff-7e4ed6a1b9b376cb82bb1a6d219be431R149